### ISSUE TITLE
misc. problems with flux resource match allocate

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-sched#466]
Encountered a few issues when testing my hwloc whitelist support for #454 

1. When there is no match, `flux resource match allocate` prints out:

```console
ahn1@b71137318c0b:/usr/src$ flux resource match allocate ./test001.yaml
JOBID                STATUS               AT                   OVERHEAD (Secs)
0                    ALLOCATED            Now                  0.0010130405426
================================================================================
MATCHED RESOURCES:
{"execution":}
```

2. with Sierra2 hwloc xml, `flux resource match allocate` prints an error message and hangs

```console
flux start -s 1
flux module remove resource
flux hwloc reload /usr/src/t/data/hwloc-data/004N/exclusive/04-brokers-sierra2/
flux module load resource
flux resource match allocate ./test001.yaml
2019-06-11T00:20:49.490515Z resource.err[0]: match_request_cb: flux_respond_error: Invalid argument
```

**Comment by dongahn** [comment]
I plan to add fixes for those as part of PR #454.

**Comment by dongahn** [comment]
Item #2 above was easy to fix. Just need to add `errno=EBUSY` when resources cannot be allocated. Changes are committed to PR #467.

Iterm #1: I reproduced it using `resource-query`.

test001.yaml:
```yaml
ahn1@b71137318c0b:/usr/src$ cat test001.yaml
version: 1
resources:
    - type: node
      count: 1
      with:
          - type: slot
            count: 1
            label: default
            with:
              - type: socket
                count: 1
                with:
                  - type: core
                    count: 1
# a comment
attributes:
  system:
    duration: 3600
tasks:
  - command: app
    slot: default
    count:
      per_slot: 1
```

```console
ahn1@b71137318c0b:/usr/src$ resource/utilities/resource-query -X /usr/src/t/data/hwloc-data/001N/amd_gpu/corona11.xml
resource-query> match allocate ./test001.yaml
      ---------socket1[1:x]
      ------corona11[1:s]
      ---cluster0[1:s]
INFO: =============================
INFO: JOBID=1
INFO: RESOURCES=ALLOCATED
INFO: SCHEDULED AT=Now
INFO: =============================
```

Apparently, matcher couldn't match the jobspec at the `core` level but still reported success. Clearly a bug. FWIW, when I normalized the hwloc with whitelist, the matching worked fine.

```console
ahn1@b71137318c0b:/usr/src$ resource/utilities/resource-query -X /usr/src/t/data/hwloc-data/001N/amd_gpu/0.xml --hwloc-whitelist=node,socket,core,gpu
INFO: Loading a matcher: CA
resource-query> match allocate ./test001.yaml
      ------------core47[1:x]
      ---------socket1[1:x]
      ------corona11[1:s]
      ---cluster0[1:s]
INFO: =============================
INFO: JOBID=1
INFO: RESOURCES=ALLOCATED
INFO: SCHEDULED AT=Now
INFO: =============================
```

**Comment by dongahn** [comment]
FYI -- the root cause of item #1 there: 

We had a partial match logic to allow a jobspec to omit the prefix of hierarchical resource requests. It turned out this partial match logic accidently and thus incorrectly attempts partial matching for non-prefix, lower-level components as well. For example,

- with resource graph: cluster->node->socket->numanode->core,
- and jobspec: node->socket->core
 
Then, this logic doesn't fail matching when it couldn't find a qualifed resource at the socket level. Notice at this level, resource graph has numanode while the jobspec looks for core. This resulted in successful match with only partially matched R ("node->socket" but no "core"!).

PR #467 fixes this by differentiating no-match case for prefix omission support vs. other types of no-match.

**Comment by dongahn** [comment]
Fixed with PR #454.
